crates/rocksdb/src/db_manager.rs

```diff
@@ -0,0 +1,330 @@
+    last_applied_opts: RocksDbOptions,
+}
+
+struct DbWatchdog {
+    manager: &'static RocksDbManager,
+    cache: Cache,
+    watchdog_rx: tokio::sync::mpsc::Receiver<WatchdogCommand>,
+    updateable_common_opts: Box<dyn Updateable<CommonOptions> + Send>,
+    current_common_opts: CommonOptions,
+    subscriptions: Vec<ConfigSubscription>,
+}
+
+impl DbWatchdog {
+    pub async fn run(
+        manager: &'static RocksDbManager,
+        watchdog_rx: tokio::sync::mpsc::Receiver<WatchdogCommand>,
+        mut updateable_common_opts: impl Updateable<CommonOptions> + Send + 'static,
+    ) -> anyhow::Result<()> {
+        let prev_opts = updateable_common_opts.load().clone();
+        let mut watchdog = Self {
+            manager,
+            cache: manager.cache.clone(),
+            watchdog_rx,
+            updateable_common_opts: Box::new(updateable_common_opts),
+            current_common_opts: prev_opts,
+            subscriptions: Vec::new(),
+        };
+
+        let shutdown_watch = cancellation_watcher();
+        tokio::pin!(shutdown_watch);
+
+        let config_watch = Configuration::watcher();
+        tokio::pin!(config_watch);
+
+        loop {
+            tokio::select! {
+                biased;
+                _ = &mut shutdown_watch => {
+                    // Shutdown requested.
+                    break;
+                }
+                Some(cmd) = watchdog.watchdog_rx.recv() => {
+                    watchdog.handle_command(cmd)
+                }
+                _ = config_watch.changed() => {
+                    watchdog.on_config_update();
+                }
+            }
+        }
+
+        watchdog.shutdown();
+        Ok(())
+    }
+
+    fn shutdown(&mut self) {
+        // Ask all databases to shutdown cleanly.
+        let start = Instant::now();
+        for v in &self.manager.dbs {
+            info!(
+                db = %v.name,
+                owner = %v.owner,
+                "Shutting down rocksdb database"
+            );
+            if let Err(e) = v.db.flush_wal(true) {
+                warn!(
+                    db = %v.name,
+                    owner = %v.owner,
+                    "Failed to flush local loglet rocksdb WAL: {}",
+                    e
+                );
+            }
+            if let Err(e) = v.db.flush_memtables_atomically(v.cfs().as_slice(), true) {
+                warn!(
+                    db = %v.name,
+                    owner = %v.owner,
+                    "Failed to flush memtables: {}",
+                    e
+                );
+            }
+            v.db.cancel_all_background_work(true);
+        }
+        info!("Clean rocksdb shutdown took {:?}", start.elapsed());
+    }
+
+    fn handle_command(&mut self, cmd: WatchdogCommand) {
+        match cmd {
+            WatchdogCommand::ResetAll => {
+                self.shutdown();
+                self.manager.dbs.clear();
+                self.subscriptions.clear();
+            }
```

Comment:
It is a bit corner casey, but I think it can happen that `manager.dbs` and `subscriptions` can go out of sync. I am thinking about the case where someone sent `ResetAll` and then someone else calls `RocksDbManager::open_db` before the `ResetAll` command has been processed. After `open_db` we have the new db in `manager.dbs` but not yet in `subscriptions`. Then we process `ResetAll` and clear `manager.dbs` and `subscriptions`. Next we process `Register` which adds the newly opened db that has already been removed to `subscriptions`.

I think the underlying problem is that we have two data structures (`manager.dbs` and `subscriptions`) that are maintained by different writers (`DbWatchdog` and thread accessing `RocksDbManager`).

Reply:
ResetAll is meant to be for tests only, I'll gate it.